enhancements/cluster-logging/elasticsearch-cert-management.md

```diff
@@ -0,0 +1,177 @@
+
+### User Stories
+
+#### As an user of EO, I do not want to have to generate and provide secrets for the Elasticsearch cluster
+
+We want to further simplify using the Elasticsearch Operator and create a single source of truth for the certs the cluster uses.
+
+#### As a consumer of EO, I want a mechanism to have certificates injected into a secret for me to continue mTLS communcation with my Elasticsearch cluster
+
+We want to continue allowing mTLS as this supports the Cluster Logging and Jaeger use cases, however they shouldn't need to do more than notify they need certs injected by EO.
+
+### Implementation Details
+
+#### Assumptions
+
+* The Elasticsearch Operator will implement this in golang (rather than the current script used in CLO)
+
+* The operator will need to be able to store generated certificates for each namespace/unique Elasticsearch cluster so that there is no security concerns of clusters cross communicating.
+
+#### Security
+
+Ensuring that signing certificates and configurations are isolated for different clusters is a must for this implementation. It will also need to be able to do this thread safe manner.
+
+### Risks and Mitigations
+
+## Not Breaking Other Operators who Depend on EO
+
+Currently there are two operators that depend on the Elasticsearch Operator and would need to ensure that their legacy functionality [read: their continuing to do cert management] is not broken. For instances where one operator is upgraded along with EO, but the other is not, we would need to provide an opt-in model for Elasticsearch to perform cert management (if we make it opt-out, we would break the legacy functionality).
+
+The most simple mechanism for this would be to gate Elasticsearch Operator performing cert management based on an annotation on the elasticsearch CR.
+
+## Design Details
+
+From a user perspective, there should be no change in any of the CRDs. The only major changes would be no longer generating secrets for EO to use, annotating their elasticsearch CR, and needing to annotate a secret (user provided) in the case they want to communicate with ES in a mTLS way.
+
+### secret annotation
+
+We will add an annotation value to check for on secrets in namespaces where an es CR exists:
+
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: fluentd
+  annotations:
+    secret.openshift.io/elasticsearch-cert-component-name: system.logging.fluentd
+type: Opaque
+data: {}
+```
+
+Where the value will be optional (defaults to the name of the secret) and corresponds to the CN used for the generated certificate.
+
+### elasticsearch cr annotation
+
+We will require an annotation value for opting in to the Elasticsearch Operator generating and managing certificates for the particular cluster:
+
+```
+apiVersion: logging.openshift.io/v1
+kind: Elasticsearch
+metadata:
+  name: elasticsearch
+  annotations:
+    logging.openshift.io/elasticsearch-cert-management: true
```

Comment:
The pattern elsewhere (e.g. cert gen service we use for metrics) to tell the service which secret will be injected in lieu of watching all secrets

Reply:
it does that because it creates certs specific for that service. this is the annotation that would be on the elasticsearch CR, which follows a similar pattern to what you're describing... we don't need to know the name of the secret to inject into because we will be responsible for creating the secret here (and given it needs to match the name of the elasticsearch CR name, we can infer it)

Reply:
> it does that because it creates certs specific for that service. this is the annotation that would be on the elasticsearch CR, which follows a similar pattern to what you're describing...

If we make providers define the name thats the trigger that indicates and sets expectations the operator is managing the content of the secret, or at least the keys it provides

> we don't need to know the name of the secret to inject into because we will be responsible for creating the secret here (and given it needs to match the name of the elasticsearch CR name, we can infer it)

This statement sounds contrary to at least what is described above which is there is some secret, presumably created by the CLO (e.g. fluentd) and not EO.  The CLO (or any client) needs to know the name of this secret because its used to mount into some other service (e.g. fluentd).  It makes more sense for the requestor to tell the provider where to inject certs, for how else would we know the name of the created secret?   This is why the cert service pattern works well for consumers.

Reply:
> This statement sounds contrary to at least what is described above which is there is some secret, presumably created by the CLO (e.g. fluentd) and not EO. The CLO (or any client) needs to know the name of this secret because its used to mount into some other service (e.g. fluentd). It makes more sense for the requestor to tell the provider where to inject certs, for how else would we know the name of the created secret? This is why the cert service pattern works well for consumers.

You had responded on the annotation for the es CR so I thought you were referring to that...

```
apiVersion: v1
kind: Secret
metadata:
  name: fluentd
  annotations:
    secret.openshift.io/elasticsearch-cert-component-name: system.logging.fluentd
type: Opaque
data: {}
```

so given the idea that we annotate the elasticsearch CR to follow the pattern of the cert signing service, how would we get the cn and the name of the secret to inject into - something like the following?
```
apiVersion: logging.openshift.io/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
  annotations:
    logging.openshift.io/elasticsearch-cert-management: true
    logging.openshift.io/client.secret/system.logging.fluentd: fluentd
```

Reply:
The existing pattern is the requester creates the secret and the service injects and knows about it because of the annotation on the service like https://github.com/openshift/cluster-logging-operator/blob/master/pkg/k8shandler/fluentd.go#L79-L96
> 
> so given the idea that we annotate the elasticsearch CR to follow the pattern of the cert signing service, how would we get the cn and the name of the secret to inject into - something like the following?
> 

Why does the requester care about the CN name?  Why cant the provider generate a CN name using a known combination of NS, ES name, secret name (e.g. openshift-logging.elasticsearch.fluentd) or similiar?  

This might be something like:
```
 apiVersion: logging.openshift.io/v1
 kind: Elasticsearch
 metadata:
   name: elasticsearch
   namespace: openshift-logging
   annotations:
      logging.openshift.io/elasticsearch-secret-name: fluentd
 ```

Maybe the requester also needs to request the roles he requires so we can properly seed the permissions: 

```
  logging.openshift.io/elasticsearch-roles: read,write
```
or as I mentioned in a previous comment we need to be explicit about the roles we give.  I presume we should not have to do this for kibana deployments because user's express they need kibana and  the certs are only needed between kibana and elasticsearch

Reply:
> Why does the requester care about the CN name? Why cant the provider generate a CN name using a known combination of NS, ES name, secret name (e.g. openshift-logging.elasticsearch.fluentd) or similiar?

this would make our implementation that much more difficult then, we would be updating and seeding the elasticsearch security config on the fly for each cluster because we've decided to allow a new configuration of DN on the client certs. For a new cluster it is already complicated but if we have to upgrade a cluster that is moving from CL/Jaeger managing certs to EO doing so and we also need to reseed their cluster... i think this is well outside the scope of what we are looking to achieve here

Reply:
> > Why does the requester care about the CN name? Why cant the provider generate a CN name using a known combination of NS, ES name, secret name (e.g. openshift-logging.elasticsearch.fluentd) or similiar?
> 
> this would make our implementation that much more difficult then, we would be updating and seeding the elasticsearch security config on the fly for each cluster because we've decided to allow a new configuration of DN on the client certs. For a new cluster it is already complicated but if we have to upgrade a cluster that is moving from CL/Jaeger managing certs to EO doing so and we also need to reseed their cluster... i think this is outside the scope of what we are looking to achieve here

I guess we need to take this in context of how far work will go with the operator.  The realistic expectation is that only jaeger and CLO will make use of cert gen meaning we can continue to rely on the static security config that exist today.  The implications are there is the possibility that either of those consumers don't specify what matches the static config and thus are broken.  That risk is likely small.  In this context the following is probably sufficient:

```apiVersion: logging.openshift.io/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
  annotations:
    logging.openshift.io/elasticsearch-client-secret/system.logging.fluentd: fluentd
```
I believe the existence of the one annotation is sufficient to indicate a consumer needs certs injected

Reply:
> I believe the existence of the one annotation is sufficient to indicate a consumer needs certs injected

Except for the case of Kibana and ES are stood up and managed by EO... we would need to know if EO is going to be responsible for creating the certs so that they can communicate with one another or if someone else is going to provide that for us. Hence the other annotation (I'm thinking about use cases where someone wants to just stand up ES and Kibana together... its not a typical use case, however it is specific for testing in the EO repo)

Reply:
>  logging.openshift.io/elasticsearch-cert-management: true

I'm missing the value here?  If I define a singular annotation is that not sufficient to indicate I want the EO to provide and manage certs? If i define nothing isn't the expectation that I am managing the certs?

Reply:
> If i define nothing isn't the expectation that I am managing the certs?

the intention of this is that eventually we always move it to where EO is creating/maintaining certs for its clusters. however, it will need to be an opt in, hence having the second annotation which will eventually be ignored/deprecated. We would still need the other annotation to denote that you need a client cert generated so you can communicate with ES.

Reply:
Tying off this discussion from our verbal chat:
* one annotation says:  I want EO to manage certs for EO and the cluster itself to communicate
* one annotation says: I want EO to give me certs so a client can talk to the cluster